server/data/healthcheck.js

```diff
@@ -1,41 +1,57 @@
 /* eslint-disable prefer-promise-reject-errors */
 const superagent = require('superagent')
+const Agent = require('agentkeepalive')
+const { HttpsAgent } = require('agentkeepalive')
 const db = require('./dataAccess/db')
 const logger = require('../../log.js')
+const config = require('../config')
 
 function dbCheck() {
   return db.queryWithoutTransaction('SELECT 1 AS ok')
 }
 
-function serviceCheck(name, url) {
-  return new Promise((resolve, reject) => {
-    superagent
-      .get(url)
-      .timeout({
-        response: 4000,
-        deadline: 4500,
-      })
-      .end((error, result) => {
-        try {
-          if (error) {
-            logger.error(error.stack, `Error calling ${name}`)
-            return reject(`${error.status} | ${error.code} | ${error.errno}`)
-          }
+const agentOptions = {
+  maxSockets: config.apis.elite2.agent.maxSockets,
+  maxFreeSockets: config.apis.elite2.agent.maxFreeSockets,
+  freeSocketTimeout: config.apis.elite2.agent.freeSocketTimeout,
+}
 
-          if (result.status === 200) {
-            return resolve('OK')
-          }
+function serviceCheckFactory(name, url) {
+  const keepaliveAgent = url.startsWith('https') ? new HttpsAgent(agentOptions) : new Agent(agentOptions)
+  logger.warn(`We have constructed a new service check here for ${url}`)
 
-          return reject(result.status)
-        } catch (apiError) {
-          logger.error(apiError.stack, `Exception calling ${name} service`)
-          return reject(apiError)
-        }
-      })
-  })
+  return () =>
+    new Promise((resolve, reject) => {
+      superagent
+        .get(url)
+        .agent(keepaliveAgent)
+        .retry(2, (err, res) => {
+          if (err) logger.info(`Retry handler found API error with ${err.code} ${err.message}`)
+          return undefined // retry handler only for logging retries, not to influence retry logic
+        })
+        .timeout({
+          response: 1000,
+          deadline: 1500,
+        })
+        .end((error, result) => {
+          try {
+            if (error) {
+              logger.error(error.stack, `Error calling ${name}`)
+              reject(error)
+            } else if (result.status === 200) {
+              resolve('OK')
+            } else {
+              reject(result.status)
+            }
+          } catch (apiError) {
```

Comment:
not quite sure what would throw this error? the code in the try seems to only log or return a promise?